mlrun/utils/http.py

```diff
@@ -0,0 +1,136 @@
+                    method_whitelist=self._get_retry_methods(retry_on_post),
+                    # we want to retry but not to raise since we do want that last response (to parse details on the
+                    # error from response body) we'll handle raising ourselves
+                    raise_on_status=False,
+                ),
+                pool_maxsize=int(config.httpdb.max_workers),
+            )
+
+            self.mount("http://", http_adapter)
+            self.mount("https://", http_adapter)
+
+    def request(self, method, url, **kwargs):
+        retry_count = 0
+        while True:
+            try:
+                response = super().request(method, url, **kwargs)
+                return response
+            except tuple(self.HTTP_RETRYABLE_EXCEPTIONS.keys()) as exc:
+                if not self.retry_on_exception:
+                    logger.warning(
+                        f"{method} {url} request failed, http retries disabled, raising exception: {exc}",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                        retry_interval=self.retry_backoff_factor,
+                        retry_count=retry_count,
+                        max_retries=self.max_retries,
+                    )
+                    raise exc
+
+                if retry_count >= self.max_retries:
+                    logger.warning(
+                        f"Maximum retries exhausted for {method} {url} request",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                        retry_interval=self.retry_backoff_factor,
+                        retry_count=retry_count,
+                        max_retries=self.max_retries,
+                    )
+                    raise exc
+
+                # only retry on exceptions with the right message
+                exception_is_retryable = any(
+                    [
+                        msg in str(exc)
+                        for msg in self.HTTP_RETRYABLE_EXCEPTIONS[type(exc)]
+                    ]
+                )
+
+                if not exception_is_retryable:
+                    logger.warning(
+                        f"{method} {url} request failed on non-retryable exception",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                    )
+                    raise exc
+
+                if self.verbose:
+                    logger.debug(
+                        f"{method} {url} request failed on retryable exception, "
+                        f"retrying in {self.retry_backoff_factor} seconds",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                        retry_interval=self.retry_backoff_factor,
+                        retry_count=retry_count,
+                        max_retries=self.max_retries,
+                    )
```

Comment:
And a fourth time here (only now it's `debug` rather than `warning` - could also use the same internal function, as long as it has a parameter saying whether to use debug or warning...